### PR TITLE
Detect match arm body without braces

### LINKED ISSUE #82524: Account for match arms statements without a block
[Given the following code](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=70547431cac2d13423237b0bf40179a6):

```rust
    match S::get(1) {
        Some(Val::Foo) => ();
        _ => (),
    }
```

The current output is:

```
error: expected one of `,`, `.`, `?`, `}`, or an operator, found `;`
  --> src/main.rs:26:29
   |
26 |         Some(Val::Foo) => ();
   |                        --   ^ expected one of `,`, `.`, `?`, `}`, or an operator
   |                        |
   |                        while parsing the `match` arm starting here

error[E0283]: type annotations needed
  --> src/main.rs:25:11
   |
4  |     fn get<K, V: Default>(_: K) -> Option<V> {
   |                  ------- required by this bound in `S::get`
...
25 |     match S::get(1) {
   |           ^^^^^^ cannot infer type for type parameter `V` declared on the associated function `get`
   |
   = note: cannot satisfy `_: Default`
help: consider specifying the type arguments in the function call
   |
25 |     match S::get::<K, V>(1) {
   |                 ^^^^^^^^
```

The first error is a generic parse error that we could account for. The second is an inference error that would be completely solved if the parse error was recovered more gracefully.

Ideally, for a generic case the output should look like:

```
error: expected one of `,`, `.`, `?`, `}`, or an operator, found `;`
  --> src/main.rs:26:29
   |
26 |         Some(Val::Foo) => ();
   |                        --   ^ expected one of `,`, `.`, `?`, `}`, or an operator
   |                        |
   |                        while parsing the `match` arm starting here
help: you likely meant to write a match arm block
   |
26 |         Some(Val::Foo) => { (); }
   |                           ^     ^
```

while for a case where a single statement is found we could also suggest:

```
error: expected one of `,`, `.`, `?`, `}`, or an operator, found `;`
  --> src/main.rs:26:29
   |
26 |         Some(Val::Foo) => ();
   |                        --   ^ expected one of `,`, `.`, `?`, `}`, or an operator
   |                        |
   |                        while parsing the `match` arm starting here
help: you might have meant to write a match arm expression separator
   |
26 |         Some(Val::Foo) => (),
   |                             ^
```

_Taken from https://twitter.com/mcclure111/status/1364998004852236289_

### PR DESCRIPTION
**Pull request by estebank**
Fix #82524.

### CODE REVIEW

File: compiler/rustc_parse/src/parser/expr.rs [hunk 1]
```diff
@@ -1973,6 +1973,102 @@ impl<'a> Parser<'a> {
...
+                    vec![
+                        (span.shrink_to_lo(), "{ ".to_string()),
+                        (span.shrink_to_hi(), " }".to_string()),
+                    ],
+                    Applicability::MachineApplicable,
+                );
+            } else {
+                err.span_suggestion(
+                    semi_sp,
+                    "use a comma to end a match arm expression",
```
Comment (JohnTitor) [inline compiler/rustc_parse/src/parser/expr.rs:2017]:
```suggestion
                    "use a comma to end a `match` arm expression",
```

File: compiler/rustc_parse/src/parser/expr.rs [hunk 2]
```diff
@@ -1973,6 +1973,102 @@ impl<'a> Parser<'a> {
...
+                    ",".to_string(),
+                    Applicability::MachineApplicable,
+                );
+            }
+            err.emit();
+            this.mk_expr_err(span)
+        };
+        // We might have either a `,` -> ;` typo, or a block without braces. We need
```
Comment (JohnTitor) [inline compiler/rustc_parse/src/parser/expr.rs:2025]:
```suggestion
        // We might have either a `,` -> `;` typo, or a block without braces. We need
```

**Review by oli-obk — COMMENTED** [review]
r=me with nits resolved

**Pull request merged**